### PR TITLE
Refactoring base vel/acc IK codes with a base class

### PR DESCRIPTION
**Pull request by robodreamer**
- all the common functions are moved to base class now
- updated velocity and acceleration IK codes so that
  they use the functions in the base class

This update was written based on Matt's initial attempt. I fixed a bug in his version and made acceleration IK use base class functions as well.

### CODE REVIEW

**Comment by robodreamer** [comment]
I have updated the code based on Matt's comments.

File: sns_ik_lib/include/sns_ik/sns_ik_base.hpp [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+   * Set the bounds on the configuration space variable. These limits match the order of
+   * the solver, for example, on a velocity solver this will set the velocity limits.
+   * Set a bound to infinity (or an arbitrarily large value) to disable it.
+   * This method will update the number of joints in the solver based on the size of qLow and qUpp
+   * Requirements: inputs must be the same size and qLow < qUpp
+   * @param qLow: lower bound on the velocity/acceleration of each joint
+   * @param qUpp: upper bound on the velocity/acceleration of each joint
+   * @return: true iff successful
+   */
+  bool setBnd(const Eigen::ArrayXd& qLow, const Eigen::ArrayXd& qUpp);
```
Comment (IanTheEngineer) [inline sns_ik_lib/include/sns_ik/sns_ik_base.hpp]: This strikes me as an unnecessary abbreviation. I'd prefer to see it named `setBounds` or even `setJointConfigBounds`

File: sns_ik_lib/include/sns_ik/sns_ik_base.hpp [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+
+  /*
+   * @return: reference to the lower bound in configuration space (pos / vel / acc)
+   */
+  const Eigen::ArrayXd& getLow() const { return qLow_; };
```
Comment (IanTheEngineer) [inline sns_ik_lib/include/sns_ik/sns_ik_base.hpp]: Again, the more descriptive names the better, ie. `getLowerBounds`

File: sns_ik_lib/include/sns_ik/sns_ik_base.hpp [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+
+  /*
+   * @return: reference to the lower bound in configuration space (pos / vel / acc)
+   */
+  const Eigen::ArrayXd& getUpp() const { return qUpp_; };
+
+  /*
+   * @return: number of joints
+   */
+  int nJnt() const { return nJnt_; }
```
Comment (IanTheEngineer) [inline sns_ik_lib/include/sns_ik/sns_ik_base.hpp]: I'd follow KDL's lead and name this [`getNrOfJoints()`](https://github.com/orocos/orocos_kinematics_dynamics/blob/4c09fb9849da8952107a5efe7aa6b2ca58c68d0e/orocos_kdl/src/chain.hpp#L71)

File: sns_ik_lib/include/sns_ik/sns_ik_base.hpp [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+   */
+  SnsIkBase(int nJnt) : nJnt_(nJnt), qLow_(nJnt), qUpp_(nJnt) {};
+
+  /*
+   * Check that qLow_ <= q <= qUpp_
+   * @param qUpp: joint acceleration to test
+   * @return: true iff qLow <= q <= qUpp
+   *          if q.empty() return false
+   */
+  bool checkBnd(const Eigen::VectorXd& q);
```
Comment (IanTheEngineer) [inline sns_ik_lib/include/sns_ik/sns_ik_base.hpp]: `checkBounds`

File: sns_ik_lib/include/sns_ik/sns_ik_base.hpp [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+
+  /*
+   * @return: reference to the lower bound in configuration space (pos / vel / acc)
+   */
+  const Eigen::ArrayXd& getLow() const { return qLow_; };
+
+  /*
+   * @return: reference to the lower bound in configuration space (pos / vel / acc)
+   */
+  const Eigen::ArrayXd& getUpp() const { return qUpp_; };
```
Comment (IanTheEngineer) [inline sns_ik_lib/include/sns_ik/sns_ik_base.hpp]: `getUpperBounds`

File: sns_ik_lib/include/sns_ik/sns_ik_base.hpp (hunk 3 above)
Comment (IanTheEngineer) [reply sns_ik_lib/include/sns_ik/sns_ik_base.hpp]: The return value should be of type `unsigned int` or a `size_t`

File: sns_ik_lib/include/sns_ik/sns_ik_base.hpp [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+   * @param[out] q: solution to the linear system
+   * @param[out] resErr: residual error in the linear system
+   * @return: Success if the solve was successful
+   */
+  SnsIkBase::ExitCode solveLinearSystem(const Eigen::MatrixXd& rhs, Eigen::VectorXd* q, double* resErr);
+
+  /*
+   * @return: rank of the matrix that is currently set in the linear solver
+   */
+  int getLinSolverRank() const { return linSolver_.rank(); }
```
Comment (IanTheEngineer) [inline sns_ik_lib/include/sns_ik/sns_ik_base.hpp]: Can this solver return a negative rank?

File: sns_ik_lib/src/sns_acc_ik_base.cpp [hunk 7]
```diff
@@ -226,40 +168,36 @@ SnsAccIkBase::ExitCode SnsAccIkBase::solve(const Eigen::MatrixXd& J, const Eigen
     // Saturate the most critical joint
     W(jntIdx, jntIdx) = 0.0;
     jointIsFree[jntIdx] = false;
-    if ((*ddq)(jntIdx) > ddqUpp_(jntIdx)) {
-      ddqNull(jntIdx) = ddqUpp_(jntIdx);
-    } else if ((*ddq)(jntIdx) < ddqLow_(jntIdx)) {
-      ddqNull(jntIdx) = ddqLow_(jntIdx);
+    if ((*ddq)(jntIdx) > (getUpp())(jntIdx)) {
```
Comment (IanTheEngineer) [inline sns_ik_lib/src/sns_acc_ik_base.cpp]: A prime example of where the naming changes that I outlined above would make this code easier to read.

**Review by IanTheEngineer — CHANGES_REQUESTED** [review]
These changes are great! I just have recommendations on naming conventions

File: sns_ik_lib/include/sns_ik/sns_ik_base.hpp (hunk 6 above)
Comment (robodreamer) [reply sns_ik_lib/include/sns_ik/sns_ik_base.hpp]: No it shouldn't. I will change it to size_t as well.

File: sns_ik_lib/src/sns_acc_ik_base.cpp (hunk 7 above)
Comment (robodreamer) [reply sns_ik_lib/src/sns_acc_ik_base.cpp]: I agree.

File: sns_ik_lib/include/sns_ik/sns_ik_base.hpp (hunk 3 above)
Comment (robodreamer) [reply sns_ik_lib/include/sns_ik/sns_ik_base.hpp]: Good suggestion.

File: sns_ik_lib/include/sns_ik/sns_ik_base.hpp (hunk 6 above)
Comment (IanTheEngineer) [reply sns_ik_lib/include/sns_ik/sns_ik_base.hpp]: in this context, i think `unsigned int` makes more sense, as it does not represent the size of a std container

**Review by IanTheEngineer — APPROVED** [review]
Perfect, just modify the last return type from `size_t` to `unsigned int` and i'll squash merge.

File: sns_ik_lib/include/sns_ik/sns_ik_base.hpp [hunk 8]
```diff
@@ -149,7 +149,7 @@ class SnsIkBase {
   /*
    * @return: rank of the matrix that is currently set in the linear solver
    */
-  size_t getLinSolverRank() const { return linSolver_.rank(); }
+  uint getLinSolverRank() const { return linSolver_.rank(); }
```
Comment (IanTheEngineer) [inline sns_ik_lib/include/sns_ik/sns_ik_base.hpp]: `uint` actually isn't a defined type in C++. This must be `unsigned int`

File: sns_ik_lib/include/sns_ik/sns_ik_base.hpp [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+   * @param[out] q: solution to the linear system
+   * @param[out] resErr: residual error in the linear system
+   * @return: Success if the solve was successful
+   */
+  SnsIkBase::ExitCode solveLinearSystem(const Eigen::MatrixXd& rhs, Eigen::VectorXd* q, double* resErr);
+
+  /*
+   * @return: rank of the matrix that is currently set in the linear solver
+   */
+  uint getLinSolverRank() const { return linSolver_.rank(); }
```
Comment (IanTheEngineer) [inline sns_ik_lib/include/sns_ik/sns_ik_base.hpp]: This needs to be `unsigned int` as `uint` isn't guaranteed to be defined.

**Review by IanTheEngineer — CHANGES_REQUESTED** [review]

**Pull request merged**